radix-engine/src/kernel/substate_io.rs

```diff
@@ -0,0 +1,678 @@
+use crate::kernel::call_frame::{
+    CallFrameDrainSubstatesError, CallFrameRemoveSubstateError, CallFrameScanKeysError,
+    CallFrameScanSortedSubstatesError, CallFrameSetSubstateError, CloseSubstateError,
+    CreateNodeError, DropNodeError, MoveModuleError, OpenSubstateError, PersistNodeError,
+    WriteSubstateError,
+};
+use crate::kernel::heap::{Heap, HeapRemoveNodeError};
+use crate::kernel::node_refs::NonGlobalNodeRefs;
+use crate::kernel::substate_locks::SubstateLocks;
+use crate::track::interface::{
+    CallbackError, NodeSubstates, StoreAccess, SubstateStore, TrackedSubstateInfo,
+};
+use radix_engine_common::prelude::{NodeId, PartitionNumber};
+use radix_engine_common::types::{SortedU16Key, SubstateKey};
+use radix_engine_common::ScryptoSbor;
+use radix_engine_interface::api::LockFlags;
+use radix_engine_interface::types::IndexedScryptoValue;
+use radix_engine_store_interface::db_key_mapper::SubstateKeyContent;
+use sbor::prelude::Box;
+use sbor::prelude::Vec;
+use sbor::rust::collections::LinkedList;
+use utils::prelude::IndexMap;
+use utils::rust::prelude::IndexSet;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum SubstateDevice {
+    Heap,
+    Store,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LockData {
+    flags: LockFlags,
+    device: SubstateDevice,
+}
+
+pub trait SubstateIOHandler<E> {
+    fn on_persist_node(&mut self, heap: &Heap, node_id: &NodeId) -> Result<(), E>;
+
+    fn on_store_access(&mut self, heap: &Heap, store_access: StoreAccess) -> Result<(), E>;
+}
+
+pub trait SubstateReadHandler {
+    type Error;
+
+    fn on_read_substate(
+        &mut self,
+        heap: &Heap,
+        value: &IndexedScryptoValue,
+        location: SubstateDevice,
+    ) -> Result<(), Self::Error>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor)]
+pub enum ProcessSubstateIOWriteError {
+    NonGlobalRefNotAllowed(NodeId),
+    PersistNodeError(PersistNodeError),
+}
+
+pub struct SubstateIOWrite {
+    pub move_nodes_from_heap: IndexSet<NodeId>,
+    pub add_non_global_refs: IndexMap<NodeId, SubstateDevice>,
+    pub remove_non_global_refs: IndexSet<NodeId>,
+}
+
+pub struct SubstateIO<'g, S: SubstateStore> {
+    pub heap: Heap,
+    pub store: &'g mut S,
+    pub non_global_node_refs: NonGlobalNodeRefs,
+    pub substate_locks: SubstateLocks<LockData>,
+}
+
+impl<'g, S: SubstateStore + 'g> SubstateIO<'g, S> {
```

Comment:
Just a thought, how about we split the trait into `ReferenceTrackedSubstateStore` and `ReferenceIgnoredSubstateStore`?